### PR TITLE
feat(orchestrator): turn a proven Workflow into a reusable skill

### PR DESCRIPTION
**Pull request by monobyte**
## Summary

- **What changed.** A Workflow that has completed a run gets a **Skill** button next to **Reflect**. It runs one user-triggered pass that reads the plan, the run digests (failures and recoveries included) and the loop's insights, and drafts a reusable `SKILL.md`. The draft is reviewed and fully editable; only what the user saves is written, into the profile's normal skills directory. Spec: `docs/features/orchestration/sero-orchestrator/specs/18-skill-extraction.md`.
- **Why.** Today a Workflow's knowledge is reusable only inside Orchestrator — the Library and the Catalog re-run the same machine. A skill teaches the method to every agent: the chat session, a subagent, a Room member, a later Workflow step. Library answers "run that workflow again"; a skill answers "do this the way that workflow proved works".

Five parts:

1. **Host seam.** New `appRuntime.skills` capability (`list` + `write`), guarded three ways, all in the host: the built-in gate (a runtime that fails it has no `host.skills` at all); a renderer-issued one-time approval (below); and host-side path resolution — the caller supplies a *name*, never a path. Replacing overwrites the file discovery resolved for that name, including a nested one, and an ambiguous name is refused. The write is atomic and hot-reloads sessions. The gate moved to a shared `capabilities/builtin-gate.ts`; persistent-sessions delegates to it. File mechanics were extracted from the skills IPC handler into `features/skills/store.ts`, so an Admin write and a runtime write share one authority — and its frontmatter is now serialized with the same `yaml` library the SDK parses with, which fixes a latent bug for the Admin editor too (a description containing a colon, a `#`, a quote or a newline used to save and then fail to load).
2. **`skill-creator` template refresh.** The bundled template was the upstream Anthropic skill: written for Claude and shipping a `.skill` packaging path Sero does not have. It is now agent-neutral and Sero-focused (where skills live, how they load, how Admin edits them), the three Python scripts are gone, and the references are de-branded. Note: `copyMissingDirs` only copies a template when the destination is absent, so existing profiles keep their copy — the extraction pass therefore treats `skill-creator` as optional.
3. **Extraction pass.** `skill-extract.ts` runs as a **read-only background agent** in the loop's workspace, so it can load `skill-creator` and read the files the plan names. It writes nothing; it returns a draft or an explicit refusal. A Workflow that teaches nothing durable produces no draft — the same no-churn guarantee reflection makes.
4. **Actions.** `extract_skill`, `save_skill`, `discard_skill_draft` on the coordinator, the tool, and (for the two that take only a loop id) the slash command. `save_skill` is refused unless the loop's draft is `pending` and its id matches the one being saved.
5. **UI.** `SkillDraftControl`: the button, the review dialog with editable name/description/body, a rename-or-replace path on a name collision, and a plain line when the pass declines. A pending draft reopens from its watched JSON artifact, so a reload never costs another pass. Saving back into the skill this Workflow already produced replaces it without a collision round trip.

### The write boundary

A plugin runtime's actions travel over the plugin's own agent tool, so a model can call them — and a skill is prompt content loaded into every later session, which would make "write a skill" a persistence path for anything that can talk to the model. Neither the plugin's state nor its artifacts can hold a secret from a model (they live in the workspace), so the enforcement sits on a channel the model does not have:

- the app calls `skills.approveSkillWrite` (renderer-only IPC) before dispatching a save, naming the pending draft and a SHA-256 of `[name, description, body, overwrite === true]`;
- the host holds that approval in **main-process memory only** — never on disk, never in the workspace, never in a tool result — expires it in two minutes, and consumes it on first use;
- the digest is structured, not concatenated, because a description may contain newlines; and it includes `overwrite`, because replacing is a separate authority: a plain Save approval cannot be spent on a write that replaces an existing skill.

A `save_skill` call with no matching approval is refused before anything touches the disk.

## Validation

- [x] `pnpm typecheck` — 24/24 packages green
- [ ] `pnpm build` (not run; no build-affecting change)
- [x] `pnpm --filter @sero/desktop test -- --run` — 2115 passed, 2 failed. Both failures are in `features/git/worktree/exec.test.ts` and are caused by this CI environment injecting `GH_TOKEN` and `GIT_CONFIG_VALUE_1` into the process env, which those tests assert is absent. They touch no code in this diff and fail the same way on an unmodified tree.
- [x] Orchestrator suite — 1218 passed, 1 skipped; desktop app-runtime + ipc + plugins suites — 448 passed. Every test is deterministic; no live model.
- [x] docs updated — user guide, workflows reference, data-model spec, specs index, and a static UX prototype in `docs/prototypes/`.

## Safety checklist

- [x] no secrets, tokens, credentials, or machine-specific private paths committed
- [x] no unnecessary `any`, `@ts-ignore`, or `@ts-expect-error`
- [x] touched source files remain under 500 LOC — `shared/types.ts` passed the limit with the new fields, so `LoopWarning` moved to its own module (490 lines now)

## Screenshots / recordings

Static prototype of all five states (control row, review, name taken, declined, after saving): `docs/prototypes/sero-workflow-skill-extraction.html`.

## Risk / rollout notes

- **breaking changes:** none. Both new `Loop` fields are optional and absent on every existing Workflow, so a Workflow that is never extracted behaves exactly as today. The skills IPC handler is a refactor over the extracted store; its only behaviour change is the safer frontmatter serialization.
- **new dependency:** `yaml` as a direct dependency of `@sero/desktop` (already present transitively — it is what the SDK's own frontmatter parser uses).
- **security or privacy impact:** this adds a way to write a file that is loaded into every agent session. Three independent guards cover it — the built-in gate, the renderer-issued content-and-authority-bound one-time approval, and host-side path resolution — and a saved skill remains an ordinary user skill under the existing visibility controls.
- **follow-up work:** extraction from a completed Room (the action shape, the draft record, the dialog and the host seam are all source-agnostic, so it is a new prompt and one more entry point); bundled `scripts/`/`references/` in a generated skill; a way for a plugin UI to navigate to another app's panel, which is why the saved skill shows its name on the button instead of linking to Admin → Skills; bringing template refreshes to existing profiles, which `copyMissingDirs` deliberately does not do today.

---
_Generated by [Claude Code](https://claude.ai/code/session_019qi4BaPyHuUnVBg58m8PGy)_

### CODE REVIEW

File: plugins/sero-orchestrator-plugin/ui/components/SkillDraftControl.tsx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+export function SkillDraftControl({
+  loop,
+  busy,
+  onDispatch,
+}: {
+  loop: Loop;
+  busy: boolean;
+  onDispatch: (params: Record<string, unknown>) => Promise<Record<string, unknown> | null>;
+}) {
```
Comment (monobyte) [inline plugins/sero-orchestrator-plugin/ui/components/SkillDraftControl.tsx:107]: Fixed in 075d062. The seven flags are now one `useReducer`: opening a draft, hitting a name collision and closing each move several fields together, so each is a single dispatched action and the intermediate states are unreachable. Typecheck and the orchestrator suite (1215 tests, including the five for this dialog) stay green.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by monobyte** [comment]
Reviewed current head `075d062`. The React Doctor finding is fixed, but I found four remaining issues.

### 1. High: the agent-facing tool bypasses the promised user review boundary

`save_skill` is exposed directly through the normal Orchestrator tool and accepts an arbitrary name, description, body and overwrite flag ([`extension/tools.ts` L287-L300](https://github.com/sero-labs/sero/blob/075d06277b48311aeaa8d5048525756b709eeecb/plugins/sero-orchestrator-plugin/extension/tools.ts#L287-L300)). The handler only checks that the loop has *some* `skillDraft`; it does not require a pending draft, a matching draft id, or a renderer/user approval ([`skill-actions.ts` L55-L86](https://github.com/sero-labs/sero/blob/075d06277b48311aeaa8d5048525756b709eeecb/plugins/sero-orchestrator-plugin/runtime/skill-actions.ts#L55-L86)).

Unless every Orchestrator tool call already has a separate mandatory user-confirmation gate, a model can therefore write arbitrary profile-global prompt content after any draft has existed, including after that draft was saved or discarded. That defeats the PR's main safety claim that nothing is written until the user reviews and saves it, and turns workspace prompt injection into a persistence path.

I would keep the actual save mutation off the model-facing tool surface, or require a short-lived, renderer-issued approval token bound to the pending draft and consume it once. The handler should also require `skillDraft.status === "pending"` and the matching draft id.

### 2. High: “Replace” does not replace valid nested skills

The conflict check correctly searches the recursively discovered skills and retains the existing `filePath` ([`skill-actions.ts` L74-L77](https://github.com/sero-labs/sero/blob/075d06277b48311aeaa8d5048525756b709eeecb/plugins/sero-orchestrator-plugin/runtime/skill-actions.ts#L74-L77)). However, after the user chooses Replace, that path is discarded and the host always writes `<skills>/<name>/SKILL.md` ([`skills.ts` L61-L73](https://github.com/sero-labs/sero/blob/075d06277b48311aeaa8d5048525756b709eeecb/apps/desktop/electron/features/apps/runtime/capabilities/skills.ts#L61-L73)).

Nested skills are explicitly supported by the shared store. For example, if `tavily-ai-skills/skills/tavily/search/SKILL.md` declares `name: search`, clicking Replace creates a second `skills/search/SKILL.md` and leaves the original untouched. The same happens whenever the directory name differs from the frontmatter name.

The host should resolve the existing name to exactly one host-discovered canonical path and overwrite that path, without accepting a caller-supplied path. If resolution is missing or ambiguous, refuse the replacement and require a rename.

### 3. Medium: editable/model-generated descriptions are not safely serialised as YAML

`serializeValue` converts strings with `String(val)`, then writes them unquoted as `key: value` ([`store.ts` L57-L70](https://github.com/sero-labs/sero/blob/075d06277b48311aeaa8d5048525756b709eeecb/apps/desktop/electron/features/skills/store.ts#L57-L70)). A perfectly natural description such as `Build recovery: use when installs fail`, or any multiline value from the two-row textarea, can produce invalid or misparsed frontmatter. The skill may save successfully but then fail to load.

Please use a real YAML serializer, or at minimum encode string scalars safely, and add round-trip tests for colons, hashes, quotes and newlines.

### 4. Low: the documented saved-skill affordance is not implemented

The spec says a saved skill gets a badge linking to Admin → Skills, and that re-extraction of a linked skill preselects overwrite. The current control only appends the name inside the extraction button ([`SkillDraftControl.tsx` L144-L153](https://github.com/sero-labs/sero/blob/075d06277b48311aeaa8d5048525756b709eeecb/plugins/sero-orchestrator-plugin/ui/components/SkillDraftControl.tsx#L144-L153)); clicking it starts another extraction, and the next save still sends `overwrite: undefined` until a collision round trip occurs.

Please either implement the linked badge/navigation and linked overwrite state described by the spec, or adjust the spec and PR description so the shipped behaviour is explicit.

The first three should be resolved before merge.

**Comment by monobyte** [comment]
All four confirmed and fixed in `06f845d`. Details per finding:

### 1. The review boundary — fixed with a channel a model does not have

You were right, and the two remedies you offered are not equivalent here: a token issued through the tool result is readable by whatever called `extract_skill`, and the plugin's own state and artifacts live inside the workspace, so nothing the runtime can hand the renderer is secret from a model. The enforcement had to sit on a channel that is renderer-only.

- New IPC `skills.approveSkillWrite` (renderer → main). The Orchestrator app calls it before dispatching a save, naming the pending draft (`<loopId>:<draftId>`) and a SHA-256 of the exact bytes it is about to send.
- The approval is held in main-process memory only (`features/skills/write-approvals.ts`) — never on disk, never in the workspace, never in a tool result. It expires in 2 minutes and is consumed on first use. A mismatched hash consumes nothing, so a wrong guess cannot burn the user's pending approval.
- `host.skills.write` refuses without a matching approval, before it resolves a path or touches the disk. Content-binding also closes the race you did not have to spell out: a model that waits for the user to approve and then calls `save_skill` with its own body fails the hash.
- `save_skill` additionally requires `skillDraft.status === "pending"` and `draftId` to match the current draft, so an already-saved or discarded draft is no longer a licence to write anything.

The action stays on the tool surface (the app dispatches through it), but it can no longer write without the person who reviewed the draft.

### 2. Replace on nested skills — fixed

The host now resolves the name against its own recursive listing and overwrites *that* file; `<skills>/<name>/SKILL.md` is used only when nothing matched. A name matching more than one skill is refused as ambiguous rather than resolved by guesswork. Test covers exactly your example — `vendor-pack/skills/search-tool/SKILL.md` declaring `name: search` — and asserts no duplicate appears at the canonical path.

### 3. YAML — fixed with the SDK's own library

`serializeFrontmatter` now uses `yaml.stringify` (added as a direct dependency of `@sero/desktop`; the SDK's `parseFrontmatter` already parses with it), so the write round-trips by construction. Round-trip test covers a colon, a `#`, quotes, an `@` lead-in and a newline, parsed back with the SDK's own `parseFrontmatter`. This fixes the Admin editor too, which shared the hand-rolled serializer.

### 4. Badge and linked overwrite — half implemented, half corrected

- Linked overwrite is implemented: saving into the name this Workflow already produced now replaces it without a collision round trip.
- The badge linking to Admin → Skills is **not** implemented, and the spec no longer claims it. There is no way for a plugin UI to navigate to another app's panel today — the only `openApp` affordance opens the plugin's own app from a notification. Adding one is a host affordance, not this spec's, so the spec and the PR description now state what ships: the saved skill name on the button, and editing in Admin → Skills. Happy to add the navigation in a follow-up if you want it.

### Validation

`pnpm typecheck` green (24/24). Orchestrator suite 1218 passed / 1 skipped, desktop app-runtime + ipc + plugins suites 448 passed. New tests: approval missing / mismatched / reused / expired, nested replace, ambiguous name, YAML round trip, stale and already-decided draft ids, and the app issuing its approval before the save.

One caveat worth stating rather than hiding: on one orchestrator-suite run I saw a single failure that I could not identify from the reporter output and could not reproduce in four subsequent full-suite runs or nine runs of the new UI test file. The new dialog test waits a real tick for the content hash, which is the timing I would suspect; I will keep watching it on CI and chase it if it reappears.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by monobyte** [comment]
Re-reviewed `06f845d`. The four previous findings are addressed well, including the renderer-only approval channel, pending draft binding, nested replacement, YAML round-tripping, and the corrected UI/spec contract.

There is one remaining blocker in the approval digest itself, and it should be the last requested change.

### High: the approval hash is ambiguous and does not bind overwrite authority

The digest is currently built by concatenating the three fields with newlines:

[`write-approvals.ts` L35-L39](https://github.com/sero-labs/sero/blob/06f845dea1abcde4452b77ae2baaeea5ae365b7b/apps/desktop/electron/features/skills/write-approvals.ts#L35-L39)

Because descriptions can now validly contain newlines, different inputs can produce the same preimage without breaking SHA-256. For example:

- approved: `description = "use when x"`, `body = "first\nsecond"`
- submitted: `description = "use when x\nfirst"`, `body = "second"`

Both hash `name\nuse when x\nfirst\nsecond`, but serialize to different SKILL.md content.

The approval also omits `overwrite`. The renderer decides that separately and sends it after approving the content ([`SkillDraftControl.tsx` L127-L138](https://github.com/sero-labs/sero/blob/06f845dea1abcde4452b77ae2baaeea5ae365b7b/plugins/sero-orchestrator-plugin/ui/components/SkillDraftControl.tsx#L127-L138)). A matching-content caller can therefore consume a normal Save approval with `overwrite: true` and replace an existing skill even though the user never clicked Replace.

Please hash a canonical structured representation that includes the write authority, for example `JSON.stringify([name, description, body, overwrite === true])`, identically in renderer and main. Add tests proving that moving a newline-separated segment across the description/body boundary and changing `overwrite` both fail approval.

Everything else in the revision looks merge-ready. Once this binding is fixed and the focused tests are green, I expect to give the PR a green light rather than start another round of discretionary nits.

**Comment by monobyte** [comment]
Both halves of that were real. Fixed in `0ddbac8`.

**Ambiguous preimage.** Your example reproduces exactly: with descriptions now legitimately multi-line, `name\ndescription\nbody` has no unique reading. Both sides now hash `JSON.stringify([name, description, body, overwrite === true])`, which escapes the separators inside each field, so one digest has one meaning.

**Replace authority.** Also correct, and the worse of the two: `overwrite` was decided by the renderer after the approval was issued, so a matching-content caller could spend a plain Save approval on a replacement. It is now inside the digest, and the UI computes it *before* approving so the approval and the dispatch cannot disagree.

Tests, all in the focused files:

- `binds each field, so a newline moved across the description/body boundary fails` — your exact pair (`"use when x"` / `"first\nsecond"` vs `"use when x\nfirst"` / `"second"`), and it then asserts the untouched approval still works for its own content, so the refusal is not just "approval was burned".
- `cannot be spent on a replace the user did not ask for` — Save approval, write with `overwrite: true`, refused.
- The UI tests now recompute the host's canonical digest with `node:crypto` and assert the exact hash the app sent for a plain save (`overwrite` false) and for a Replace (`overwrite` true), so a drift between the renderer and main fails a test rather than a save.

Validation: `pnpm typecheck` 24/24, orchestrator 1218 passed / 1 skipped, desktop app-runtime + ipc 315 passed. The one unreproducible failure I flagged last round has not recurred in any run since.

Spec updated with why the tuple is structured and why `overwrite` is inside it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by monobyte** [comment]
Re-reviewed `0ddbac8`. The final approval-binding issue is fixed correctly.

- Renderer and main now hash the same canonical structured tuple: `[name, description, body, overwrite === true]`.
- Newlines inside editable fields are escaped by the JSON framing, so the description/body boundary is unambiguous.
- Replace authority is decided before approval and included in the digest, so a plain Save approval cannot be spent on an overwrite.
- Focused tests cover the exact boundary-shift case, an unapproved overwrite, renderer/main digest agreement, and both plain-save and replace paths.
- React Doctor is green on this head. The reported local typecheck and focused suites are also green; GitHub's Test workflow remains skipped while the PR is draft.

I found no substantive regressions in the revision and no merge blockers remain. **Green light from review: mark the PR ready and merge once the repository's normal required checks are satisfied.**
